In [1]:
import pandas as pd
import numpy as np
import ast
from datetime import datetime

načtení původní tabulky

In [2]:
trafficDelaysDf = pd.read_csv('C:/Users/verav/Documents/Projekt/Python_projekt/plynulostVSE.csv')

In [3]:
trafficDf = trafficDelaysDf.loc[:,['properties.uuid','properties.street']]
trafficDf.columns = trafficDf.columns.str.replace('properties.','')

In [4]:
trafficDf.head()

uuid                street
0  1363839630  » Centrum / Bohunice
1  1388730276                   NaN
2  1451737922               U Dráhy
3  1388985166              Na Hrázi
4  1440208034                   NaN

In [7]:
streets2Df = pd.read_csv('missing_streets2.csv', encoding = 'UTF-8') #načtení prvního souboru doplněných ulic
streets2Df = streets2Df.drop(['Unnamed: 0', 'Unnamed: 0.1','missingStreet','firstCoord'], axis = 1)
streets2Df.head()

uuid                         streets2
0  1388730276                           Labská
1  1440208034                         Okrouhlá
2  1370182069                           Labská
3  1415306095  Starý Lískovec,okres Brno-město
4  1415188035                         Okrouhlá

zbavení se nedoplněných ulic (městská část,okres) v Streets2

In [8]:
def nullMissing(vstup):#funkce na náhradu špatného názvu prázdným řetězcem
    if 'okres' in vstup:
        new = vstup.replace(vstup,"")
    else:
        new = vstup
    return new

In [9]:
streets2Df['streets2'].apply(nullMissing)[0:5]

0      Labská
1    Okrouhlá
2      Labská
3            
4    Okrouhlá
Name: streets2, dtype: object

In [10]:
streets2Df['streets2'] = streets2Df['streets2'].apply(nullMissing)

In [11]:
streetsMerge1Df = trafficDf.merge(streets2Df,how ='left', on = 'uuid') #propojení tabulek podle uuid

In [12]:
streetsMerge1Df.head(10)

uuid                street   streets2
0  1363839630  » Centrum / Bohunice        NaN
1  1388730276                   NaN     Labská
2  1451737922               U Dráhy        NaN
3  1388985166              Na Hrázi        NaN
4  1440208034                   NaN   Okrouhlá
5  1370182069                   NaN     Labská
6  1415306095                   NaN           
7  1363840440  » Centrum / Bohunice        NaN
8  1415188035                   NaN   Okrouhlá
9  1531577943                   NaN  Jihlavská

In [14]:
streetsMerge1Df['streets2'] = streetsMerge1Df['streets2'].fillna("")
streetsMerge1Df['street'] = streetsMerge1Df['street'].fillna("")

In [15]:
streetsMerge1Df.head(10)

uuid                street   streets2
0  1363839630  » Centrum / Bohunice           
1  1388730276                           Labská
2  1451737922               U Dráhy           
3  1388985166              Na Hrázi           
4  1440208034                         Okrouhlá
5  1370182069                           Labská
6  1415306095                                 
7  1363840440  » Centrum / Bohunice           
8  1415188035                         Okrouhlá
9  1531577943                        Jihlavská

doplnění názvů do sloupce street ze sloupce streets2

In [16]:
street = []
for i in streetsMerge1Df.index:
    if len(streetsMerge1Df['street'][i]) == 0:
        street.append(streetsMerge1Df['streets2'][i])
    else:
        street.append(streetsMerge1Df['street'][i])

In [17]:
streetsMerge1Df['street'] = street
streetsMerge1Df.head()

uuid                street  streets2
0  1363839630  » Centrum / Bohunice          
1  1388730276                Labská    Labská
2  1451737922               U Dráhy          
3  1388985166              Na Hrázi          
4  1440208034              Okrouhlá  Okrouhlá

kontrola počtu chybějících názvů ulic

In [18]:
pocetMiss1 = []
for i in streetsMerge1Df.index:
    if len(streetsMerge1Df['street'][i]) == 0:
        pocetMiss1.append(streetsMerge1Df['street'][i])

In [19]:
len(pocetMiss1)

885

2.kolo doplňování názvů (z poslední souřadnice)

In [20]:
streets3Df = pd.read_csv('missing_streets3.csv', encoding = 'UTF-8')
streets3Df = streets3Df[['uuid','street3']]
streets3Df.head()

uuid                            street3
0   1415306095                           Okrouhlá
1   1337977851                         Charbulova
2   1415181365  Moravany u Brna,okres Brno-venkov
3   1531488523                        Bubeníčkova
4   1440329274  Moravany u Brna,okres Brno-venkov
5   1364049280  Moravany u Brna,okres Brno-venkov
6   1531481363  Moravany u Brna,okres Brno-venkov
7   1370400299                          Mrštíkova
8   1411773276  Nebovidy u Brna,okres Brno-venkov
9   1403581057                           Brněnská
10  1392955330        Žabovřesky,okres Brno-město
11  1499609402  Střelice u Brna,okres Brno-venkov
12  1423009730                          Mrštíkova
13  1448061796                           Okrouhlá
14  1499783774                        Bubeníčkova
15  1431005108        Žabovřesky,okres Brno-město
16  1426861420        Žabovřesky,okres Brno-město
17  1434372218                             Hlavní
18  1458262396                           Křivánky
19  1407727051                             Hlavní

In [22]:
streetsMerge2Df = streetsMerge1Df.merge(streets3Df,how ='left', on = 'uuid') #propojení tabulek podle uuid

In [23]:
streetsMerge2Df.head()

uuid                street   streets2   street3
0   1363839630  » Centrum / Bohunice                  NaN
1   1388730276                Labská     Labská       NaN
2   1451737922               U Dráhy                  NaN
3   1388985166              Na Hrázi                  NaN
4   1440208034              Okrouhlá   Okrouhlá       NaN
5   1370182069                Labská     Labská       NaN
6   1415306095                                   Okrouhlá
7   1363840440  » Centrum / Bohunice                  NaN
8   1415188035              Okrouhlá   Okrouhlá       NaN
9   1531577943             Jihlavská  Jihlavská       NaN
10  1369227948   Mikuláškovo náměstí                  NaN
11  1337986071              Vltavská   Vltavská       NaN
12  1388808276              Na Hrázi                  NaN
13  1434832415                    D1                  NaN
14  1338028001               U Dráhy                  NaN
15  1451855772           Zapletalova                  NaN
16  1369238908            Opletalova                  NaN
17  1370398359                Labská     Labská       NaN
18  1546901953              Zvonařka                  NaN
19  1531563033             Lužánecká                  NaN

In [24]:
streetsMerge2Df['street3']= streetsMerge2Df['street3'].fillna("")

In [25]:
streetsMerge2Df

uuid                street  streets2 street3
0       1363839630  » Centrum / Bohunice                  
1       1388730276                Labská    Labská        
2       1451737922               U Dráhy                  
3       1388985166              Na Hrázi                  
4       1440208034              Okrouhlá  Okrouhlá        
...            ...                   ...       ...     ...
116704  1503042209              Lesnická                  
116705   435163611                  Oblá                  
116706   649666312               Koliště                  
116707   649898012                   374                  
116708  1503213619               Lidická                  

[116709 rows x 4 columns]

In [31]:
streetList2 = []
for i in streetsMerge2Df.index:
    if len(streetsMerge2Df['street'][i]) == 0:
        new = streetsMerge2Df['street3'][i]
        streetList2.append(new)
    else:
        streetList2.append(streetsMerge2Df['street'][i])

In [40]:
streetsMerge2Df['street'] = streetList2
streetsMerge2Df.head()

uuid                          street      streets2     street3
0   1363839630            » Centrum / Bohunice                          
1   1388730276                          Labská        Labská            
2   1451737922                         U Dráhy                          
3   1388985166                        Na Hrázi                          
4   1440208034                        Okrouhlá      Okrouhlá            
5   1370182069                          Labská        Labská            
6   1415306095                        Okrouhlá                  Okrouhlá
7   1363840440            » Centrum / Bohunice                          
8   1415188035                        Okrouhlá      Okrouhlá            
9   1531577943                       Jihlavská     Jihlavská            
10  1369227948             Mikuláškovo náměstí                          
11  1337986071                        Vltavská      Vltavská            
12  1388808276                        Na Hrázi                          
13  1434832415                              D1                          
14  1338028001                         U Dráhy                          
15  1451855772                     Zapletalova                          
16  1369238908                      Opletalova                          
17  1370398359                          Labská        Labská            
18  1546901953                        Zvonařka                          
19  1531563033                       Lužánecká                          
20  1415312495                          Kalová                          
21  1337977851                      Charbulova                Charbulova
22  1415313065                          Průhon                          
23  1381967467                     Křídlovická                          
24  1364069340                      Ondráčkova                          
25  1369049068                          Plotní                          
26  1453563552                          Uhelná                          
27  1440151424                     Zapletalova                          
28  1451755692                          Spojka                          
29  1415125595                        Musilova                          
30  1451843832                          Veveří                          
31  1531579413                     Zahradnická                          
32  1531586043                      Bělohorská                          
33  1337834751                       Tomáškova                          
34  1369046038                           Polní                          
35  1337832531                     Skopalíkova                          
36  1364052330                         Viniční                          
37  1531494483                       Mrštíkova     Mrštíkova            
38  1415306885                        Hoštická                          
39  1369043128                            Vrbí          Vrbí            
40  1451754642                       Tomáškova                          
41  1451748222                      Klusáčkova                          
42  1388802356                  » Vev. Bitýška                          
43  1388729906                          Krátká                          
44  1369038388  » Praha / Bratislava / Olomouc                          
45  1440317684                         Dornych                          
46  1363841460                             384                          
47  1415191715                       Vomáčkova                          
48  1388978396                        Grmelova                          
49  1370402259                             373           373            
50  1368955768                         Viniční                          
51  1356333081                       Heršpická                          
52  1337738091                          Spěšná                          
53  1388971556                    Svatoplukova  Svatoplukova        

In [46]:
streetsMerge2Df['street']= streetsMerge2Df['street'].fillna("") #převede automaticky vše na string

In [73]:
chybi = []

In [74]:
for i in streetsMerge2Df.index:
    if 'okres' in streetsMerge2Df['street'][i]:
        chybi.append(streetsMerge2Df['street'][i])

In [75]:
len(chybi)

462

In [76]:
streetsMerge2Df['street'] = streetsMerge2Df['street'].str.strip()

z předchozího postupu zbývá nedoplněných ulic cca 462, doplněno alespoň městskou částí a názvem obce...

In [79]:
finalDfStreet = streetsMerge2Df[['uuid','street']]
len(finalDfStreet)

116709

In [ ]:
finalDfStreet.to_csv('finalDfStreet.csv', encoding = 'utf-8', index = False)